File: src/services/clips-history.ts
import { readFile, writeFile, mkdir, rm } from "fs/promises";
import { existsSync } from "fs";
import { basename, join } from "path";
import { v4 as uuidv4 } from "uuid";
import { paths } from "../config/paths.js";
import { writeFileAtomic } from "../utils/atomic-file.js";
import { withFileLock, FileLockError } from "../utils/mutation-lock.js";
import { childLogger } from "../utils/logger.js";
import { sliceTranscript, sliceWords } from "../utils/transcript.js";
import { isDemoMode, demoClips } from "../ui/demo-fixtures.js";
import type { BatchClipsResult, ClipHistoryEntry, Format, WordTimestamp } from "../models/index.js";

const log = childLogger("clips-history");

type BatchResultRow = BatchClipsResult["results"][number];

interface BatchRecordContext {
  sourceVideo: string;
  transcriptWords?: WordTimestamp[] | null;
  defaultCaptionStyle?: string;
  defaultCropStrategy?: string;
  defaultFormat?: Format;
  contentTypeFor?: (start: number, end: number) => string | undefined;
}

export interface BatchClipSpec {
  start_second: number;
  end_second: number;
  keep_segments?: Array<{ start: number; end: number }>;
}

export interface BatchRecipeContext {
  transcriptWords?: WordTimestamp[] | null;
  logoPath?: string | null;
  outroPath?: string | null;
  introPath?: string | null;
  cleanFillers?: boolean;
  clipSpecs?: BatchClipSpec[];
}

export type HistoryReadErrorCode = "HISTORY_UNREADABLE" | "HISTORY_INVALID_JSON" | "HISTORY_INVALID_SHAPE";

/** The history file exists but cannot be used; mutations abort without writing. */
export class HistoryReadError extends Error {
  readonly code: HistoryReadErrorCode;
  readonly path: string;
  constructor(code: HistoryReadErrorCode, message: string, path: string) {
    super(message);
    this.name = "HistoryReadError";
    this.code = code;
    this.path = path;
  }
}

export { FileLockError };

function isErrno(err: unknown, code: string): boolean {
  return typeof err === "object" && err !== null && (err as NodeJS.ErrnoException).code === code;
}

// Minimum compatible structure: an array of objects, each with a non-empty
// string id. Every reader and writer already depends on id; no other field is
// required so legacy entries keep working.
function validateShape(data: unknown, path: string): ClipHistoryEntry[] {
  if (!Array.isArray(data)) {
    throw new HistoryReadError(
      "HISTORY_INVALID_SHAPE",
      `History file ${path} must contain a JSON array of clip entries. No changes were written.`,
      path,
    );
  }
  data.forEach((entry, index) => {
    const id = entry && typeof entry === "object" && !Array.isArray(entry) ? (entry as { id?: unknown }).id : undefined;
    if (typeof id !== "string" || !id) {
      throw new HistoryReadError(
        "HISTORY_INVALID_SHAPE",
        `History file ${path} entry ${index} is not a clip record with a string id. No changes were written.`,
        path,
      );
    }
  });
  return data as ClipHistoryEntry[];
}

/**
 * Read the history file for a mutation. A missing file yields an empty list
 * and `raw: null`; any other failure throws HistoryReadError so the caller
 * cannot overwrite a file it could not read.
 */
export async function readHistoryStrict(
  path: string,
): Promise<{ entries: ClipHistoryEntry[]; raw: string | null }> {
  let raw: string;
  try {
    raw = await readFile(path, "utf-8");
  } catch (err) {
    if (isErrno(err, "ENOENT")) return { entries: [], raw: null };
    throw new HistoryReadError(
      "HISTORY_UNREADABLE",
      `History file ${path} could not be read (${(err as Error).message}). No changes were written.`,
      path,
    );
  }
  const text = raw.replace(/^﻿/, "");
  let data: unknown;
  try {
    data = JSON.parse(text);
  } catch (err) {
    throw new HistoryReadError(
      "HISTORY_INVALID_JSON",
      `History file ${path} is not valid JSON (${(err as Error).message}). No changes were written. ` +
        "Restore it from a backup or repair the file, then retry.",
      path,
    );
  }
  return { entries: validateShape(data, path), raw: text };
}

// Resolve a full id or an unambiguous ≥4-char prefix within a given list.
function resolveIdIn(entries: ClipHistoryEntry[], idOrPrefix: string): string | null {
  if (!idOrPrefix) return null;
  if (entries.some((e) => e.id === idOrPrefix)) return idOrPrefix;
  if (idOrPrefix.length < 4) return null;
  const matches = entries.filter((e) => e.id.startsWith(idOrPrefix));
  return matches.length === 1 ? matches[0].id : null;
}

export class ClipsHistory {
  private historyPath = paths.clipsHistory;
  // Serializes this instance's own read-modify-write cycles so concurrent HTTP
  // requests queue instead of spinning on the file lock. Cross-process and
  // cross-language safety (other Studio instances, the Python CLI) comes from
  // the shared mutation lock held inside mutate().
  private writeChain: Promise<unknown> = Promise.resolve();
  private syncing = new Set<string>();

  private async ensureDir() {
    if (!existsSync(paths.history)) {
      await mkdir(paths.history, { recursive: true });
    }
  }

  // Lenient read for listing and lookups: a missing, unreadable, or invalid
  // file reads as empty so the UI keeps working, and the failure is logged.
  // Mutations never use this path; see mutate().
  async load(): Promise<ClipHistoryEntry[]> {
    if (isDemoMode()) return demoClips();
    try {
      return (await readHistoryStrict(this.historyPath)).entries;
    } catch (err) {
      log.warn(`clip history unavailable: ${(err as Error).message}`);
      return [];
    }
  }

  // Run fresh read → mutate → atomic save as one critical section under the
  // shared lock, queued behind any in-flight mutation of this instance. The
  // callback returns the value the caller wants back. Demo fixtures are
  // read-only: the callback runs against a fresh copy and nothing persists.
  private mutate<T>(fn: (entries: ClipHistoryEntry[]) => T | Promise<T>): Promise<T> {
    const run = this.writeChain.then(async () => {
      if (isDemoMode()) return fn(demoClips());
      await this.ensureDir();
      return withFileLock(
        this.historyPath,
        async () => {
          const { entries } = await readHistoryStrict(this.historyPath);
          const before = JSON.stringify(entries);
          const result = await fn(entries);
          // Rewrite only when the content changed: a lookup that finds nothing
          // must not touch the file, and a missing file stays missing.
          if (JSON.stringify(entries) !== before) {
            await writeFileAtomic(this.historyPath, JSON.stringify(entries, null, 2));
          }
          return result;
        },
        { tool: "clipperz-studio" },
      );
    });
    this.writeChain = run.then(() => undefined, () => undefined);
    return run;
  }

  async record(entry: Omit<ClipHistoryEntry, "id" | "created_at">): Promise<ClipHistoryEntry> {
    const full: ClipHistoryEntry = {
      ...entry,
      id: uuidv4(),
      created_at: new Date().toISOString(),
    };
    await this.mutate((entries) => {
      entries.push(full);
    });
    void this.syncToCloud(full);
    return full;
  }

  /**
   * Mirror a rendered clip to the workspace, for signed-in users.
   *
   * Deliberately not awaited and unable to throw: a clip that rendered
   * successfully must be recorded locally whether or not a server was reachable.
   * The local history file remains the source of truth; this is a copy.
   *
   * Clips that fail to sync are left marked so a later sweep can backfill them —
   * the performance model wants the whole history, not the part that happened to
   * have a working network.
   */
  private async syncToCloud(entry: ClipHistoryEntry): Promise<void> {
    // record() starts this in the background, so `podcli sync` can reach the
    // same entry while it is still in flight and register the clip twice.
    if (this.syncing.has(entry.id)) return;
    this.syncing.add(entry.id);
    try {
      const cloud = await import("./podcli-cloud.js");
      if (!(await cloud.signedIn())) return;

      const source = entry.source_video;
      if (!source) return;

      const clipId = entry.cloud_id ?? (await cloud.registerClip({
        sourceHash: await cloud.sourceHash(source),
        episodeTitle: basename(source),
        title: entry.title,
        startSecond: entry.start_second,
        endSecond: entry.end_second,
        durationSec: entry.duration,
        contentType: entry.content_type,
        captionStyle: entry.caption_style,
        aspectRatio: entry.format,
        transcriptSlice: entry.transcript_slice,
      }))?.id;
      if (!clipId) return;

      await this.update(entry.id, { cloud_id: clipId });

      // Metadata alone leaves a share link with nothing to play, so the
      // rendered file follows it. Uploaded once: the server keeps the first
      // copy and answers `unchanged` after that.
      let hasVideo = entry.cloud_video_uploaded === true;
      // A rendered file that no longer exists locally can never be uploaded.
      // There is nothing left to do for it, and reporting it as failed on every
      // run is the unfixable number this file already refuses to print.
      const uploadable = !hasVideo && existsSync(entry.output_path);

      if (uploadable) {
        hasVideo = await cloud.uploadClipVideo(clipId, entry.output_path);
        if (hasVideo) await this.update(entry.id, { cloud_video_uploaded: true });
      }

      // Synchronised means the clip is watchable, not merely described: an
      // upload that failed while still reporting success is how `podcli sync`
      // exits happy with every share link playing nothing.
      await this.update(entry.id, {
        cloud_synced: hasVideo || !existsSync(entry.output_path),
      });
    } catch {
      await this.update(entry.id, { cloud_synced: false }).catch(() => {});
    } finally {
      this.syncing.delete(entry.id);
    }
  }

  // Persist every successful row of a batch render. Single source of truth for
  // turning backend batch results into history entries — callers used to inline
  // this loop, drifting on defaults and on which fields got recorded.
  async recordBatchResults(
    results: BatchResultRow[] | undefined,
    ctx: BatchRecordContext,
  ): Promise<ClipHistoryEntry[]> {
    if (!results) return [];
    const recorded: ClipHistoryEntry[] = [];
    for (const r of results) {
      if (r.status !== "success" || !r.output_path) continue;
      const start = r.start_second || 0;
      const end = r.end_second || 0;
      recorded.push(
        await this.record({
          source_video: ctx.sourceVideo,
          start_second: start,
          end_second: end,
          caption_style: r.caption_style || ctx.defaultCaptionStyle || "hormozi",
          crop_strategy: r.crop_strategy || ctx.defaultCropStrategy || "speaker",
          format: r.format || ctx.defaultFormat || "vertical",
          title: r.title || "clip",
          output_path: r.output_path,
          file_size_mb: r.file_size_mb || 0,
          duration: r.duration || 0,
          content_type: ctx.contentTypeFor?.(start, end),
          transcript_slice: sliceTranscript(ctx.transcriptWords, start, end),
        }),
      );
    }
    return recorded;
  }

  async persistBatchRecipes(
    rows: BatchResultRow[] | undefined,
    recorded: ClipHistoryEntry[],
    ctx: BatchRecipeContext,
  ): Promise<void> {
    if (!rows?.length || !recorded.length) return;
    let recordedIdx = 0;
    for (const row of rows) {
      if (row.status !== "success" || !row.output_path) continue;
      const rec = recorded[recordedIdx++];
      if (!rec) continue;
      const spec =
        typeof row.clip_index === "number" ? ctx.clipSpecs?.[row.clip_index] : undefined;
      await this.persistClipRecipe(rec, {
        transcriptWords: ctx.transcriptWords,
        logoPath: ctx.logoPath,
        outroPath: ctx.outroPath,
        introPath: ctx.introPath,
        cleanFillers: ctx.cleanFillers,
        keepSegments: spec?.keep_segments,
      });
    }
  }

  async persistClipRecipe(
    rec: ClipHistoryEntry,
    ctx: {
      transcriptWords?: WordTimestamp[] | null;
      logoPath?: string | null;
      outroPath?: string | null;
      introPath?: string | null;
      cleanFillers?: boolean;
      foregroundFraming?: import('../services/foreground-framing.js').ForegroundFraming | null;
      captionPosition?: string;
      captionFontScale?: number;
      logoPosition?: string;
      keepSegments?: Array<{ start: number; end: number }>;
    },
  ): Promise<void> {
    const words = sliceWords(ctx.transcriptWords ?? [], rec.start_second, rec.end_second);
    await this.saveWords(rec.id, words);
    await this.saveRecipe(rec.id, {
      foreground_framing: ctx.foregroundFraming ?? null,
      caption_position: ctx.captionPosition || "auto",
      caption_font_scale: ctx.captionFontScale || 100,
      logo_position: ctx.logoPosition || "top-left",
      caption_style: rec.caption_style,
      crop_strategy: rec.crop_strategy,
      format: rec.format || "vertical",
      logo_path: ctx.logoPath ?? rec.logo_path ?? null,
      outro_path: ctx.outroPath ?? rec.outro_path ?? null,
      intro_path: ctx.introPath ?? rec.intro_path ?? null,
      clean_fillers: ctx.cleanFillers ?? false,
      transcript_words: words,
      ...(ctx.keepSegments?.length && { keep_segments: ctx.keepSegments }),
    });
    if (ctx.keepSegments?.length) {
      await this.update(rec.id, { keep_segments: ctx.keepSegments });
    }
  }

  /**
   * Check if a clip with the same source, time range, and style already exists.
   * Uses basename matching for source video and ±2s tolerance on time range.
   */
  async findDuplicate(
    sourceVideo: string,
    startSecond: number,
    endSecond: number,
    captionStyle: string,
    cropStrategy: string,
    format: string = "vertical"
  ): Promise<ClipHistoryEntry | null> {
    const entries = await this.load();
    const srcName = basename(sourceVideo);

    return (
      entries.find((e) => {
        if (basename(e.source_video) !== srcName) return false;
        if (e.caption_style !== captionStyle) return false;
        if (e.crop_strategy !== cropStrategy) return false;
        if ((e.format || "vertical") !== format) return false;
        if (Math.abs(e.start_second - startSecond) > 2) return false;
        if (Math.abs(e.end_second - endSecond) > 2) return false;
        // Check output still exists
        return existsSync(e.output_path);
      }) || null
    );
  }

  async list(limit = 50): Promise<ClipHistoryEntry[]> {
    const entries = await this.load();
    return entries.slice(-limit).reverse();
  }

  // Exact match only — REST routes feed req.params.id straight in, so a loose
  // prefix could target the wrong clip (and an empty prefix the first one).
  async findById(id: string): Promise<ClipHistoryEntry | undefined> {
    if (!id) return undefined;
    const entries = await this.load();
    return entries.find((e) => e.id === id);
  }

  // Resolve a full id or an unambiguous ≥4-char prefix to a full id. For the
  // human-facing MCP tool, where typing a short prefix is convenient.
  async resolveId(idOrPrefix: string): Promise<string | null> {
    return resolveIdIn(await this.load(), idOrPrefix);
  }

  async update(id: string, patch: Partial<ClipHistoryEntry>): Promise<ClipHistoryEntry | null> {
    if (!id) return null;
    const changed = await this.mutate((entries) => {
      const e = entries.find((x) => x.id === id);
      if (!e) return null;
      const before = e.title;
      Object.assign(e, patch);
      return { entry: e, previousTitle: before };
    });
    if (!changed) return null;

    // A human rewriting a generated title is the clearest taste signal Clipperz
    // gets — it says what the model produced and what a person preferred
    // instead. Reported only when the title actually changed, so the sync
    // bookkeeping in syncToCloud can't trigger it.
    if (patch.title !== undefined && patch.title !== changed.previousTitle) {
      void this.reportEvent(changed.entry, "title_edited", changed.previousTitle, patch.title);
    }
    return changed.entry;
  }

  /** Best-effort; never blocks or fails the edit that produced it. */
  private async reportEvent(
    entry: ClipHistoryEntry,
    kind: "title_edited" | "discarded" | "thumbnail_regenerated",
    before?: string,
    after?: string,
  ): Promise<void> {
    if (!entry.cloud_id) return;
    try {
      const cloud = await import("./podcli-cloud.js");
      if (!(await cloud.signedIn())) return;
      await cloud.logClipEvent(entry.cloud_id, kind, before, after);
    } catch {
      // The signal is nice to have, not worth surfacing an error over.
    }
  }

  /**
   * Push clips that never reached the workspace.
   *
   * Covers two cases that both matter: a render that happened while the network
   * was down, and — more importantly — everything rendered *before* the user
   * subscribed. A new Pro user should start with their back catalogue behind the
   * performance model, not an empty history.
   */
  async backfillCloud(limit = 200): Promise<{ synced: number; failed: number }> {
    const cloud = await import("./podcli-cloud.js");
    if (!(await cloud.signedIn())) return { synced: 0, failed: 0 };

    // A clip whose source video has been moved or deleted can never be hashed,
    // so it can never sync. Skipping it keeps `podcli sync` quiet; counting it
    // as a failure would report the same unfixable number on every run until
    // people stopped reading the output.
    const pending = (await this.load())
      .filter((e) => e.source_video && existsSync(e.source_video))
      .filter((e) => !e.cloud_id || !e.cloud_video_uploaded)
      .slice(0, limit);

    let synced = 0;
    let failed = 0;
    for (const entry of pending) {
      try {
        await this.syncToCloud(entry);
        const after = await this.findById(entry.id);
        if (after?.cloud_synced) synced++;
        else failed++;
      } catch {
        failed++;
      }
    }
    return { synced, failed };
  }

  // Remove a clip and the artifacts Clipperz rendered for it (output video,
  // word/recipe/reframe sidecars, thumbnail dir). The source video is never touched.
  // Accepts a full id or an unambiguous prefix (MCP convenience); the prefix is
  // resolved inside the critical section so a concurrent change cannot redirect it.
  async remove(idOrPrefix: string): Promise<ClipHistoryEntry | null> {
    if (!idOrPrefix) return null;
    // Demo entries are read-only fixtures — never delete their (shipped) artifacts.
    if (isDemoMode()) {
      const id = await this.resolveId(idOrPrefix);
      return id ? (await this.findById(id)) ?? null : null;
    }
    const entry = await this.mutate((entries) => {
      const id = resolveIdIn(entries, idOrPrefix);
      if (!id) return null;
      const idx = entries.findIndex((e) => e.id === id);
      if (idx < 0) return null;
      return entries.splice(idx, 1)[0];
    });
    if (!entry) return null;

    const artifacts = [
      this.wordsPath(entry.id),
      this.recipePath(entry.id),
      this.reframePath(entry.id),
      entry.output_path,
    ];
    await Promise.all(
      artifacts.map((p) => (p ? rm(p, { force: true }) : Promise.resolve())),
    );
    await rm(join(paths.output, "thumbnails", entry.id), { recursive: true, force: true });
    return entry;
  }

  async getBySource(videoPath: string): Promise<ClipHistoryEntry[]> {
    const entries = await this.load();
    const srcName = basename(videoPath);
    return entries.filter((e) => basename(e.source_video) === srcName).reverse();
  }

  // Word timings are kept in a sidecar (not in clips.json) so re-rendering a
  // clip can re-burn captions without bloating the history file.
  private wordsPath(id: string): string {
    return join(paths.history, "words", `${id}.json`);
  }

  async saveWords(id: string, words: unknown[]): Promise<void> {
    if (!words || words.length === 0) return;
    await mkdir(join(paths.history, "words"), { recursive: true });
    await writeFile(this.wordsPath(id), JSON.stringify(words), "utf-8");
  }

  async loadWords(id: string): Promise<unknown[]> {
    try {
      return JSON.parse(await readFile(this.wordsPath(id), "utf-8"));
    } catch {
      return [];
    }
  }

  // Full render recipe (logo/outro/captions/fillers/segments/words) so a clip
  // can be re-rendered faithfully — e.g. after a manual reframe.
  private recipePath(id: string): string {
    return join(paths.history, "recipes", `${id}.json`);
  }

  async saveRecipe(id: string, recipe: Record<string, unknown>): Promise<void> {
    await mkdir(join(paths.history, "recipes"), { recursive: true });
    await writeFile(this.recipePath(id), JSON.stringify(recipe), "utf-8");
  }

  async loadRecipe(id: string): Promise<Record<string, unknown> | null> {
    try {
      return JSON.parse(await readFile(this.recipePath(id), "utf-8"));
    } catch {
      return null;
    }
  }

  // Reframe editor state (keyframes + trim) so reopening shows prior edits.
  private reframePath(id: string): string {
    return join(paths.history, "reframe", `${id}.json`);
  }

  async saveReframe(id: string, state: Record<string, unknown>): Promise<void> {
    await mkdir(join(paths.history, "reframe"), { recursive: true });
    await writeFile(this.reframePath(id), JSON.stringify(state), "utf-8");
  }

  async loadReframe(id: string): Promise<Record<string, unknown> | null> {
    try {
      return JSON.parse(await readFile(this.reframePath(id), "utf-8"));
    } catch {
      return null;
    }
  }
}

File: backend/services/clips_history.py
"""
Clip history — read/write .podcli/history/clips.json.

Shared on-disk contract with the TypeScript ClipsHistory service
(src/services/clips-history.ts). Both languages read and write the same
file; this module is the Python writer used by the CLI.

Entry shape (all fields beyond the core render record are optional):
  id, source_video, start_second, end_second, caption_style, crop_strategy,
  logo_path?, title, output_path, file_size_mb, duration, created_at,
  content_type?, transcript_slice?, youtube_video_id?, metrics?,
  generated_titles?, description?, tags?, hashtags?

metrics? = {views?, retention?, ctr?, impressions?, fetched_at?}  (Phase 2)

Rules
- Writes are read-modify-write and MUST preserve unknown keys, so fields
  written by the other language (e.g. Phase 2 metrics) are never clobbered.
- Every mutation goes through ``mutate_clips_history``: it holds the shared
  cross-process lock (services/mutation_lock.py, same protocol as the TS side)
  from the fresh read and id resolution through the atomic replacement.
- A missing file initializes to an empty list. Invalid JSON, an invalid shape,
  or an unreadable file aborts the mutation without changing a byte.
"""

import json
import os
import shutil
import sys
import time
import uuid
from typing import Callable, Optional, TypeVar

from config.paths import paths
from services.mutation_lock import FileLockError, file_lock

T = TypeVar("T")


class ClipsHistoryError(RuntimeError):
    """Base class for surfaced history failures (lock or read)."""


class HistoryReadError(ClipsHistoryError):
    """The history file exists but cannot be used. ``code`` is one of
    HISTORY_UNREADABLE, HISTORY_INVALID_JSON, HISTORY_INVALID_SHAPE."""

    def __init__(self, code: str, message: str, path: str):
        super().__init__(message)
        self.code = code
        self.path = path


class HistoryLockError(ClipsHistoryError):
    def __init__(self, cause: FileLockError):
        super().__init__(str(cause))
        self.code = cause.code
        self.lock_path = cause.lock_path
        self.owner = cause.owner


def _history_path() -> str:
    return paths["clipsHistory"]


def _validate_shape(data: object, path: str) -> list[dict]:
    """Minimum compatible structure: a list of objects, each with a non-empty
    string ``id``. Every existing reader and writer already depends on ``id``;
    no other field is required so legacy entries keep working."""
    if not isinstance(data, list):
        raise HistoryReadError(
            "HISTORY_INVALID_SHAPE",
            f"History file {path} must contain a JSON array of clip entries. No changes were written.",
            path,
        )
    for index, entry in enumerate(data):
        if not isinstance(entry, dict) or not isinstance(entry.get("id"), str) or not entry["id"]:
            raise HistoryReadError(
                "HISTORY_INVALID_SHAPE",
                f"History file {path} entry {index} is not a clip record with a string id. No changes were written.",
                path,
            )
    return data


def read_clips_history_strict(path: Optional[str] = None) -> tuple[list[dict], Optional[str]]:
    """Return (entries, raw text). A missing file yields ([], None). Any other
    failure raises HistoryReadError so a mutation cannot overwrite it."""
    path = path or _history_path()
    try:
        with open(path, encoding="utf-8") as f:
            raw = f.read()
    except FileNotFoundError:
        return [], None
    except OSError as exc:
        raise HistoryReadError(
            "HISTORY_UNREADABLE", f"History file {path} could not be read ({exc}). No changes were written.", path
        ) from exc
    text = raw.lstrip("﻿")
    try:
        data = json.loads(text)
    except ValueError as exc:
        raise HistoryReadError(
            "HISTORY_INVALID_JSON",
            f"History file {path} is not valid JSON ({exc}). No changes were written. "
            "Restore it from a backup or repair the file, then retry.",
            path,
        ) from exc
    return _validate_shape(data, path), text


def load_clips_history() -> list[dict]:
    """Load all clip entries for reading. Returns [] if the file is missing;
    also [] (with a warning on stderr) if it is unreadable or invalid, so
    listing keeps working. Mutations never use this lenient path."""
    try:
        entries, _ = read_clips_history_strict()
        return entries
    except HistoryReadError as exc:
        print(f"  ! clip history unavailable: {exc}", file=sys.stderr)
        return []


def _serialize(entries: list[dict]) -> str:
    return json.dumps(entries, indent=2, ensure_ascii=False)


_REPLACE_RETRY_S = 1.0
_REPLACE_RETRY_STEP_S = 0.025


def _write_entries_locked(path: str, text: str) -> None:
    """Atomic replacement: temp file then os.replace. On failure the original
    file is untouched and the temp file is removed.

    Windows refuses to replace a file another process has open without
    FILE_SHARE_DELETE (a lenient reader in either language can hold it for
    milliseconds), so the replace retries briefly before the error surfaces.
    """
    os.makedirs(os.path.dirname(path), exist_ok=True)
    tmp = f"{path}.{os.getpid()}.{uuid.uuid4().hex[:8]}.tmp"
    try:
        with open(tmp, "w", encoding="utf-8") as f:
            f.write(text)
        deadline = time.monotonic() + _REPLACE_RETRY_S
        while True:
            try:
                os.replace(tmp, path)
                break
            except PermissionError:
                if time.monotonic() >= deadline:
                    raise
                time.sleep(_REPLACE_RETRY_STEP_S)
    except BaseException:
        try:
            os.remove(tmp)
        except OSError:
            pass
        raise


def mutate_clips_history(fn: Callable[[list[dict]], T], timeout_ms: Optional[int] = None) -> T:
    """Run ``fn(entries)`` as one locked read-modify-write cycle.

    ``fn`` mutates the list in place and returns whatever the caller needs.
    The file is rewritten only when the content changed, so a no-op mutation
    leaves the bytes alone.
    """
    path = _history_path()
    os.makedirs(os.path.dirname(path), exist_ok=True)
    try:
        with file_lock(path, timeout_ms=timeout_ms, tool="clipperz-cli"):
            entries, _raw = read_clips_history_strict(path)
            before = json.dumps(entries, ensure_ascii=False)
            result = fn(entries)
            # Rewrite only when the content changed: a lookup that finds nothing
            # must not touch the file, and a missing file stays missing.
            if json.dumps(entries, ensure_ascii=False) != before:
                _write_entries_locked(path, _serialize(entries))
            return result
    except FileLockError as exc:
        raise HistoryLockError(exc) from exc


def list_clips(limit: int = 50) -> list[dict]:
    """Most recent clips first."""
    entries = load_clips_history()
    return entries[-limit:][::-1]


def get_clips_by_source(video_path: str) -> list[dict]:
    """All clips from a source video (basename match), newest first."""
    target = os.path.basename(video_path)
    entries = load_clips_history()
    return [e for e in entries if os.path.basename(e.get("source_video", "")) == target][::-1]


def _find_in(entries: list[dict], clip_id: str) -> Optional[dict]:
    """Exact id match, falling back to an unambiguous prefix match."""
    if not clip_id:
        return None
    for e in entries:
        if e.get("id") == clip_id:
            return e
    prefix_matches = [e for e in entries if str(e.get("id", "")).startswith(clip_id)]
    return prefix_matches[0] if len(prefix_matches) == 1 else None


def find_clip(clip_id: str) -> Optional[dict]:
    """Find by exact id, falling back to an unambiguous prefix match."""
    return _find_in(load_clips_history(), clip_id)


def _clip_sidecar_paths(clip_id: str) -> list[str]:
    """Per-clip sidecar files (words/recipe/reframe) and the thumbnail dir."""
    history_dir = os.path.dirname(_history_path())
    return [
        os.path.join(history_dir, "words", f"{clip_id}.json"),
        os.path.join(history_dir, "recipes", f"{clip_id}.json"),
        os.path.join(history_dir, "reframe", f"{clip_id}.json"),
    ]


def delete_clip(clip_id: str) -> Optional[dict]:
    """Remove a clip from history along with its rendered output and sidecars.

    Returns the removed entry, or None if no clip matched. The source video is
    never touched, only artifacts Clipperz rendered for this clip. The id is
    resolved inside the lock so a concurrent edit cannot target a stale list.
    """
    if not clip_id:
        return None

    def remove(entries: list[dict]) -> Optional[dict]:
        target = _find_in(entries, clip_id)
        if target is None:
            return None
        entries.remove(target)
        return target

    target = mutate_clips_history(remove)
    if target is None:
        return None
    full_id = str(target.get("id"))

    artifacts = list(_clip_sidecar_paths(full_id))
    output_path = target.get("output_path")
    if output_path:
        artifacts.append(output_path)
    for path in artifacts:
        try:
            if path and os.path.isfile(path):
                os.remove(path)
        except OSError:
            pass

    thumb_dir = os.path.join(paths["output"], "thumbnails", full_id)
    try:
        if os.path.isdir(thumb_dir):
            shutil.rmtree(thumb_dir)
    except OSError:
        pass
    return target


def update_clip(clip_id: str, **fields) -> Optional[dict]:
    """Update a clip in place, preserving all unknown keys. Returns the updated entry or None.

    Only keys with non-None values are applied, so callers can pass optional
    edits without overwriting existing data with None. The id is resolved
    inside the lock; a clip deleted meanwhile is not resurrected.
    """
    if not clip_id:
        return None

    def apply(entries: list[dict]) -> Optional[dict]:
        target = _find_in(entries, clip_id)
        if target is None:
            return None
        for key, value in fields.items():
            if value is not None:
                target[key] = value
        return target

    return mutate_clips_history(apply)
